scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/cipher/CbcBlockCipher.java

```diff
@@ -70,46 +76,68 @@
 	 */
 	public static byte[] decrypt(CipherSuite suite, SecretKey key, SecretKey macKey, byte[] additionalData,
 			byte[] ciphertext) throws GeneralSecurityException {
-
 		/*
-		 * See http://tools.ietf.org/html/rfc5246#section-6.2.3.2 for
-		 * explanation
+		 * See http://tools.ietf.org/html/rfc5246#section-6.2.3.2 for explanation
 		 */
-		DatagramReader reader = new DatagramReader(ciphertext);
-		byte[] iv = reader.readBytes(suite.getRecordIvLength());
+		// extend/oversize the plaintext for MAC compensation and 256 padding checks 
+		byte[] plaintextOversized = new byte[ciphertext.length + Math.max(suite.getMacMessageBlockLength(), 256)];
+		int ivlength = suite.getRecordIvLength();
 		Cipher blockCipher = suite.getThreadLocalCipher();
-		blockCipher.init(Cipher.DECRYPT_MODE,
-				key,
-				new IvParameterSpec(iv));
-		byte[] plaintext = blockCipher.doFinal(reader.readBytesLeft());
-		// last byte contains padding length
+		blockCipher.init(Cipher.DECRYPT_MODE, key, new IvParameterSpec(ciphertext, 0, ivlength));
+		int plaintextLength = blockCipher.doFinal(ciphertext, ivlength, ciphertext.length - ivlength,
+				plaintextOversized);
+		// fill up plaintext for MAC compensation
+		System.arraycopy(FILLUP, 0, plaintextOversized, plaintextLength, suite.getMacMessageBlockLength());
```

Comment:
Is it normal that we initialize : 
`byte[] plaintextOversized = new byte[ciphertext.length + Math.max(suite.getMacMessageBlockLength(), 256)];`
And here we just use `suite.getMacMessageBlockLength()`

Is `plaintextLength == ciphertext.length` ?

Reply:
> Is plaintextLength == ciphertext.length ?

No, the ciphertext contains also the nonce so plaintext.lenght < ciphertext.length.
The array is a little more oversized. 
The `FILLUP` is filled in after the plaintext for the message digest compensation (Mac Block Length).
For the padding check, the bytes are anyway ignored (256).

Reply:
Just to mention:
If `max(plen, 256)` is used for the number of compares, the oversize will be the mac block size.